### PR TITLE
fix(core): correct order in ContentChildren query result

### LINKED ISSUE #16568: Incorrect order in ContentChildren query result
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report => search github for a similar issue or PR before submitting
```

**Current behavior**
The `@ContentChildren` query can return results in the incorrect order when querying for directives. The particular issue reported here occurs when queried directive instances "sit" on `<ng-template>`.

**Expected behavior**

The `@ContentChildren` query should return results in the expected order, regardless of directives "siting" on `<ng-template>` or not.

**Minimal reproduction of the problem with instructions**

https://plnkr.co/edit/KPz00lezleunuqt71VBn?p=preview

Check the console that prints out query result.
Observed log: `ngTabSet ts1 ta2`
Expected log: `ngTabSet ts1 ta1`

* **Angular version:** 2.0.X

4.1.0

* **Browser:** 
all

### PR DESCRIPTION
**Pull request by marclaval**
Fixes #16568

### CODE REVIEW

File: packages/core/test/linker/query_integration_spec.ts [hunk 1]
```diff
@@ -89,6 +90,20 @@ export function main() {
...
 
+      it('should contain the first content child when target is on <ng-template> with embedded view (issue #16568)',
+         () => {
+           const template =
+               '<div directive-needs-content-child><ng-template text="foo" [ngIf]="true"><div text="bar"></div></ng-template></div>' +
+               '<needs-content-child #q><ng-template text="foo" [ngIf]="true"><div text="bar"></div></ng-template></needs-content-child>';
+           const view = createTestCmp(MyComp0, template);
+           view.detectChanges();
+           const q: NeedsContentChild = view.debugElement.children[1].references !['q'];
+           expect(q.logs).toEqual([['setter', 'foo'], ['init', 'foo'], ['check', 'foo']]);
```
Comment (tbosch) [inline packages/core/test/linker/query_integration_spec.ts]: Could you change this test to not use/check the lifecycle hooks, but rather just check the content of `child.text` has the right value? (in this case it should be `foo`).

This makes the test more readable (and the `DirectiveNeedsContentChild` as well).

File: packages/core/src/view/query.ts [hunk 2]
```diff
@@ -138,8 +138,13 @@ function calcQueryValues(
     if (nodeDef.flags & NodeFlags.TypeElement && nodeDef.element !.template &&
         (nodeDef.element !.template !.nodeMatchedQueries & queryDef.filterId) ===
             queryDef.filterId) {
-      // check embedded views that were attached at the place of their template.
+      // check embedded views that were attached at the place of their template ...
```
Comment (tbosch) [inline packages/core/src/view/query.ts]: Move this comment one line down.

File: packages/core/src/view/query.ts [hunk 3]
```diff
@@ -138,8 +138,13 @@ function calcQueryValues(
...
       const elementData = asElementData(view, i);
+      // but process children first if some match the query (see issue #16568)
```
Comment (tbosch) [inline packages/core/src/view/query.ts]: Change `children` to `child nodes (i.e. directives in this case)`

**Comment by mlaval** [comment]
@tbosch PR updated and ready for another review

**Review by tbosch — APPROVED** [review]

**Pull request closed without merging**